hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/parse/QueryConverter.java

```diff
@@ -53,9 +54,12 @@
 /**
  * Converts a parse tree into a relational tree.
  */
+@SuppressWarnings("CheckStyle")
 public class QueryConverter {
     public static final SqlToRelConverter.Config CONFIG;
 
+    public static final HepProgram HEP_CALC_UNION_REWRITER_PROGRAM;
```

Comment:
I'm afraid it will not work:

> Note that the structure of a program is immutable, but the planner uses it as read/write during planning, so a program can only be in use by a single planner at a time

But maybe we could have a pool of programs, so we do not have to create it from scratch for each query.

Reply:
Nevertheless, I think this is a good direction of pre-creating and not recreating costly objects.

Reply:
Aha, fixed.

Reply:
![Uploading image.png…]()

Reply:
In Calcite 1.30 `HepPlanner` was made thread-safe, docs are lying.

![image](https://github.com/hazelcast/hazelcast/assets/10107557/78c39f5d-04f7-4fb8-a330-8dbf2599ba4e)

Reply:
Calcite javadoc is outdated: https://github.com/apache/calcite/commit/0065d7c179b98698f018f83b0af0845a6698fc54